recipe/thirdparty/providers/oauth2_impl.go

```diff
@@ -0,0 +1,242 @@
+		PKCECodeVerifier:   pkceCodeVerifier,
+	}, nil
+}
+
+func (config *typeCombinedOAuth2Config) ExchangeAuthCodeForOAuthTokens(redirectURIInfo tpmodels.TypeRedirectURIInfo, userContext supertokens.UserContext) (tpmodels.TypeOAuthTokens, error) {
+	config.discoverEndpoints()
+
+	tokenAPIURL := config.TokenEndpoint
+	accessTokenAPIParams := map[string]interface{}{
+		"client_id":    config.ClientID,
+		"redirect_uri": redirectURIInfo.RedirectURIOnProviderDashboard,
+		"code":         redirectURIInfo.RedirectURIQueryParams["code"].(string),
+		"grant_type":   "authorization_code",
+	}
+	if config.ClientSecret != "" {
+		accessTokenAPIParams["client_secret"] = config.ClientSecret
+	}
+	if redirectURIInfo.PKCECodeVerifier != nil {
+		accessTokenAPIParams["code_verifier"] = *redirectURIInfo.PKCECodeVerifier
+	}
+
+	for k, v := range config.TokenParams {
+		if v == nil {
+			delete(accessTokenAPIParams, k)
+		} else {
+			accessTokenAPIParams[k] = v
+		}
+	}
+
+	/* Transformation needed for dev keys BEGIN */
+	if isUsingDevelopmentClientId(config.ClientID) {
+		accessTokenAPIParams["client_id"] = getActualClientIdFromDevelopmentClientId(config.ClientID)
+		accessTokenAPIParams["redirect_uri"] = DevOauthRedirectUrl
+	}
+	/* Transformation needed for dev keys END */
+
+	oAuthTokens, err := doPostRequest(tokenAPIURL, accessTokenAPIParams, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return oAuthTokens, nil
+}
+
+func (config *typeCombinedOAuth2Config) GetUserInfo(oAuthTokens tpmodels.TypeOAuthTokens, userContext supertokens.UserContext) (tpmodels.TypeUserInfo, error) {
+	config.discoverEndpoints()
+
+	accessToken, accessTokenOk := oAuthTokens["access_token"].(string)
+	idToken, idTokenOk := oAuthTokens["id_token"].(string)
+
+	rawUserInfoFromProvider := tpmodels.TypeRawUserInfoFromProvider{}
+
+	if idTokenOk && config.JwksURI != "" {
+		claims := jwt.MapClaims{}
+		jwksURL := config.JwksURI
+		jwks, err := getJWKSFromURL(jwksURL)
+		if err != nil {
+			return tpmodels.TypeUserInfo{}, err
+		}
+		token, err := jwt.ParseWithClaims(idToken, claims, jwks.Keyfunc)
+		if err != nil {
+			return tpmodels.TypeUserInfo{}, err
+		}
+
+		if !token.Valid {
+			return tpmodels.TypeUserInfo{}, errors.New("invalid id_token supplied")
+		}
+		rawUserInfoFromProvider.FromIdToken = map[string]interface{}(claims)
+	}
+	if accessTokenOk && config.UserInfoEndpoint != "" {
+		headers := map[string]string{
+			"Authorization": "Bearer " + accessToken,
+		}
+		userInfoFromAccessToken, err := doGetRequest(config.UserInfoEndpoint, nil, headers)
+		rawUserInfoFromProvider.FromAccessToken = userInfoFromAccessToken.(map[string]interface{})
+
+		if err != nil {
+			return tpmodels.TypeUserInfo{}, err
+		}
+	}
+
+	userInfoResult, err := config.getSupertokensUserInfoResultFromRawUserInfo(rawUserInfoFromProvider)
+	if err != nil {
+		return tpmodels.TypeUserInfo{}, err
+	}
+
+	return tpmodels.TypeUserInfo{
+		ThirdPartyUserId:        userInfoResult.ThirdPartyUserId,
```

Comment:
you should add the tenantId here like `tenantId|userId` - we need to decide what delimiter to use.

Reply:
Done